### PR TITLE
Add ome-files-py build

### PR DESCRIPTION
**Pull request by simleo**
Adds https://github.com/ome/ome-files-py to the superbuild.

To test, run the superbuild with `-Dbuild-packages=ome-files-py` and check that ome-files-py gets built and installed to `stage/lib/python2.7/site-packages` (currently you should see only `ome_files.so` and the egg-info file). Also check that the unit tests are run.

**Note:** at the time of opening, ome/ome-files-py#2 hasn't been merged yet. If testing locally using `git_dir`, clone the `builtin_ext` branch from my fork.

### CODE REVIEW

**Comment by sbesson** [comment]
@simleo: very nice. Is there some addition to the Jenkins scripts that needs to be added (maybe limited to some OS first?) in order to have this included in the daily CI builds or is it premature? /cc @rleigh-codelibre

**Comment by rleigh-codelibre** [comment]
Yes, replace ome-files with ome-files-py in `scripts/jenkins-build*` so it's used by default.

**Comment by simleo** [comment]
Updated description to mention the test step (added in 36ba53f) and add a note on forks/branches.

File: packages/ome-files-py/superbuild.cmake [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  SOURCE_DIR "${EP_SOURCE_DIR}"
+  BINARY_DIR "${EP_BINARY_DIR}"
+  CONFIGURE_COMMAND ""
+  BUILD_COMMAND ${CMAKE_COMMAND}
+    "-DSOURCE_DIR:PATH=${EP_SOURCE_DIR}"
+    "-DBUILD_DIR:PATH=${EP_BINARY_DIR}"
+    "-DCONFIG:INTERNAL=$<CONFIG>"
+    "-DEP_SCRIPT_CONFIG:FILEPATH=${EP_SCRIPT_CONFIG}"
+    -P "${CMAKE_CURRENT_LIST_DIR}/build.cmake"
+  INSTALL_DIR ""
```
Comment (rleigh-codelibre) [inline packages/ome-files-py/superbuild.cmake]: `INSTALL_DIR` could be placed with `SOURCE_DIR` and `BINARY_DIR` at the top (for consistency with the other scripts)

**Comment by rleigh-codelibre** [comment]
This looks really nice.  Thanks for battling with cmake!

If you wanted to make the change to `scripts/jenkins-build` and `scripts/jenkins-build.bat`, this will test it on all our supported platforms.  If it turns out it needs work on some platforms, we can always move it to the "breaking" build while we fix it up.

**Comment by simleo** [comment]
Changed `scripts/jenkins-build`. As discussed with @rleigh-codelibre, we're leaving `scripts/jenkins-build.bat` alone for now as the Python extension is not going to link in the current state.

**Comment by rleigh-codelibre** [comment]
[https://ci.openmicroscopy.org/view/Files/job/OME-FILES-CPP-DEV-merge-superbuild/BUILD_TYPE=Release,node=cowfish/494/console](CentOS7 superbuild) - C++11 failure
[https://ci.openmicroscopy.org/view/Files/job/OME-FILES-CPP-DEV-merge/BUILD_TYPE=Debug,node=perch/445/console](MacOS X build) - can't find boost headers
[https://ci.openmicroscopy.org/view/Files/job/OME-FILES-CPP-DEV-merge/BUILD_TYPE=Release,node=snipe/445/console](MacOS X debug build) - C++11 failure
[https://ci.openmicroscopy.org/view/Files/job/OME-FILES-CPP-DEV-merge/BUILD_TYPE=Debug,node=snipe/445/console](FreeBSD build) - C++11 failure
[https://ci.openmicroscopy.org/view/Files/job/OME-FILES-CPP-DEV-merge/BUILD_TYPE=Release,node=perch/445/console](FreeBSD debug build) - can't find boost headers

The C++11 errors should be fixed once the module is built with `-std=c++11` or `-std=c++14`.  Note that the CMake builds all try `-std=c++14` before falling back to `-std=c++11`, and if that could be done here as well, it would be good for consistency.

The missing headers looks like it's not looking in `/usr/local/include`.  How do you tell distutils the include and library search paths for compiling and linking?  On both MacOS X and FreeBSD, this is not on the default compiler search path, so needs specifying somehow.  What's the standard way to do this?

**Comment by sbesson** [comment]
All builds are now green with the latest state of this PR. One question in the context of he testing of this PR and https://github.com/ome/ome-files-py/pull/2 is the following: how to consume `ome-files-py` using the superbuild artifacts and what environment variables should be set: `PYTHONPATH`? `PATH`? `{LD_}LIBRARY_PATH`?

**Comment by simleo** [comment]
`LD_LIBRARY_PATH` (or `DYLD_LIBRARY_PATH` on the Mac) is needed so that dynamic libraries the extension depends upon (unless installed into a standard path) can be found. PYTHONPATH (unless the extension is copied to the standard Python package dir) are needed so that Python finds the extension itself. On the Mac, I've had success with this:

```
curl -o ome-files-platform-bundle-0.2.3-29-g59acd58-Release-MacOSX10.12.2-x86_64-b449.tar.xz https://ci.openmicroscopy.org/view/CPP/job/OME-FILES-CPP-DEV-merge/BUILD_TYPE=Release,node=snipe/lastSuccessfulBuild/artifact/artefacts/binaries/ome-files-platform-bundle-0.2.3-29-g59acd58-Release-MacOSX10.12.2-x86_64-b449.tar.xz
tar xf ome-files-platform-bundle-0.2.3-29-g59acd58-Release-MacOSX10.12.2-x86_64-b449.tar.xz
cd ome-files-platform-bundle-0.2.3-29-g59acd58-Release-MacOSX10.12.2-x86_64-b449
export DYLD_LIBRARY_PATH=${PWD}/lib
export PYTHONPATH=${PWD}/lib/python2.7/site-packages
```

To use the above version you need to have dependencies other than the OME libraries already installed on your machine. I've also tried the standalone version but when I tried to import the module from Python I had a runtime problem with libtiff:

```
ImportError: dlopen(/Users/simleo/sandbox/unpack/ome-files-standalone-bundle-0.2.3-29-g705ffb1-Release-MacOSX10.12.2-x86_64-b501/lib/python2.7/site-packages/ome_files.so, 2): Symbol not found: __cg_TIFFClientOpen
  Referenced from: /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
  Expected in: /Users/simleo/sandbox/unpack/ome-files-standalone-bundle-0.2.3-29-g705ffb1-Release-MacOSX10.12.2-x86_64-b501/lib/libTIFF.dylib
 in /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
```

**Comment by sbesson** [comment]
See https://github.com/ome/ome-files-py/pull/2#issuecomment-274780309 for the functionality review. The superbuild addition seems to be working with 2 caveats:

- as noted in https://github.com/ome/ome-files-py/pull/2#issuecomment-274780309, the Python tests fail under certain conditions likely due to a misconfigured environment during the test execution
- since the inclusion of this PR, the builds have been failing intermittently when building `ome-files-py` (see the recent history of https://ci.openmicroscopy.org/job/OME-FILES-CPP-DEV-merge-superbuild/ or https://ci.openmicroscopy.org/job/OME-FILES-CPP-DEV-breaking/) with errors of type
  ```
  00:35:43 /opt/hudson/workspace/OME-FILES-CPP-DEV-breaking/BUILD_TYPE/Debug/node/cowfish/build/stage/include/ome/common/log.h:55:30: fatal error: boost/log/core.hpp: No such file or directory
  00:35:43  #include <boost/log/core.hpp>
  ```

  So far, the solution has been to clean the workspace or run the build with PURGE_CACHE set to `build/all` and we could use that as a workaround but that is something we might want to address and @rleigh-codelibre might have insights on the nature of the issue.

**Comment by rleigh-codelibre** [comment]
@simleo `DYLD_LIBRARY_PATH` should not be used on MacOS.  Use `DYLD_FALLBACK_LIBRARY_PATH` instead.  This is due to the odd way the Mach-O linker works; setting the former overrides the use of system libraries while the latter does not, and MacOS provides a libtiff which doesn't contain the complete libtiff API--this is where the `Symbol not found: __cg_TIFFClientOpen` comes from.  Don't know why they provide an incompatible proprietary libtiff API, but that's the reason why we always specify the fallback library path in the docs.

**Comment by rleigh-codelibre** [comment]
@sbesson Intermittent errors are likely a caching issue.  Try cleaning the workspace, then building, then building again.  This will test clean vs cached builds.  It was all working for months; this should not have changed that, unless the ome-files-py is making assumptions it should not.  Are the boost headers present in the cache dir?  How about the stage dir?  Where is the ome-files-py build looking for them?  Could the ome-files-py package be adjusted to look in both places.

**Comment by sbesson** [comment]
@rleigh-codelibre thanks, this makes sense. In the interim, the builds have been adjusted to unconditionally force the rebuild of the cache but we should review how headers are detected. See https://trello.com/c/SXLIplhS/59-ome-files-py-intermittent-build-failures for the corresponding card.

**Comment by simleo** [comment]
@rleigh-codelibre `packages/ome-files-py/build.cmake` is looking for headers in `${OME_EP_INCLUDE_DIR}` and `/usr/local/include`. Would adding `${OME_EP_BUILD_CACHE}/include` help?

**Comment by rleigh-codelibre** [comment]
@simleo I suspect this would resolve the problem.  We do this for the cmake builds:

```
ExternalProjectHelpers.cmake:  list(APPEND CMAKE_PREFIX_PATH "${OME_EP_BUILD_CACHE}")
ExternalProjectHelpers.cmake:  list(APPEND CMAKE_LIBRARY_PATH "${OME_EP_BUILD_CACHE}/lib")
ExternalProjectHelpers.cmake:  list(APPEND CMAKE_PROGRAM_PATH "${OME_EP_BUILD_CACHE}/bin")
```

So making the python build use this location as well would resolve the problem I think.

**Pull request merged**